mongoimport/common.go

```diff
@@ -399,28 +587,130 @@ func validateFields(fields []string) error {
 		if strings.Contains(field, "..") {
 			return fmt.Errorf("field '%v' cannot contain consecutive '.' characters", field)
 		}
-		// NOTE: since fields is sorted, this check ensures that no field
-		// is incompatible with another one that occurs further down the list.
-		// meant to prevent cases where we have fields like "a" and "a.c"
-		for _, latterField := range fieldsCopy[index+1:] {
-			// NOTE: this means we will not support imports that have fields that
-			// include e.g. a, a.b
-			if strings.HasPrefix(latterField, field+".") {
-				return fmt.Errorf("fields '%v' and '%v' are incompatible", field, latterField)
+
+		if index+1 < len(fields) {
+			// For each field except the last, we check that the next field does not
+			// contain the current field as a prefix. Since fields is sorted, this checks
+			// case (3) and case (4).
+			nextField := fields[index+1].field
+
+			if strings.HasPrefix(nextField, field+".") {
+				return fmt.Errorf("fields '%v' and '%v' are incompatible", field, nextField)
+			}
+
+			// This checks case (2).
+			if field == nextField {
+				return fmt.Errorf("fields cannot be identical: '%v' and '%v'", field, nextField)
+			}
+		}
+
+		if useArrayIndexFields {
+			var nextFieldInfo FieldInfo
+			if index+1 < len(fields) {
+				nextFieldInfo = fields[index+1]
+			}
+			// This checks cases (5), (6), (7), and (8). This is achieved by comparing each field against the
+			// next field in the sorted list.
+			err := validateFieldsUsingArrayIndexes(fieldInfo, nextFieldInfo, index, len(fields))
```

Comment:
Maybe rename `validateFieldsForArrayIndexes` as a given field may or may not have them?